[macOS] Content under left obscured content inset area is clipped when pinch-zooming

whsieh opened the pull request:
#### 83e10b359df5a0647ebb2cd5a9e597617a8973ce
<pre>
[macOS] Content under left obscured content inset area is clipped when pinch-zooming
<a href="https://bugs.webkit.org/show_bug.cgi?id=293600">https://bugs.webkit.org/show_bug.cgi?id=293600</a>
<a href="https://rdar.apple.com/151854427">rdar://151854427</a>

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

When applying transient zoom (i.e. during pinch zooming on macOS) while the left obscured content
inset is nonzero, web content underneath the obscured content inset area is clipped by the main
frame&apos;s clip layer, until the transient zoom finishes and the clip layer size and position is
updated to include the obscured content inset area.

This happens because transient zoom currently works by applying a transform to only the layer for
page scaling (i.e. the `RenderView`&apos;s primary layer); however, since the page is contained within a
clip layer whose bounds don&apos;t change while in transient zoom state (during pinch zoom), any content
that would normally show up in the obscured inset area is clipped.

To fix this, we add logic to temporarily reposition and inflate the size of the clip layer (as well
as the scrolled contents layer underneath) while the user is in transient zoom, such that the clip
layer no longer clips any part of the page that falls inside a content inset area. See below for
more details.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::clipLayer const):

Add a helper method to grab the frame view&apos;s render layer compositor&apos;s top-level clip layer.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::scaledRectAtOrigin):

Add a helper method that scales the given `FloatRect`, relative to the given `origin` point.

* Source/WebCore/platform/graphics/GeometryUtilities.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::mainFrameClipLayerID const):
(WebKit::RemoteLayerTreeTransaction::setMainFrameClipLayerID):

Add plumbing to provide the main frame clip layer ID in remote layer tree transactions, so that we
can obtain the clip layer in the UI-side remote drawing area proxy below.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::adjustTransientZoom):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/DrawingAreaProxy.h:
(WebKit::DrawingAreaProxy::adjustTransientZoom):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:

Rename the existing `m_transientZoomOrigin` to `m_transientZoomOriginInLayerForPageScale`, to
differentiate it from the new `m_transientZoomOriginInVisibleRect` (which is used to resize and
reposition the clip layer).

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):

Obtain the layer IDs of the main frame clip layer, as well as the scrolled contents layer.

(WebKit::fillFowardsAnimationWithKeyPathAndValue):
(WebKit::transientZoomTransformOverrideAnimation):
(WebKit::transientSizeAnimation):
(WebKit::transientPositionAnimation):

Add more helper functions here to create size (`bounds.size`) and position animations, which we
apply to the clipping and scrolled contents layers below.

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::applyTransientZoomToLayer):

Implement the main fix here:

1.  Compute the scaled frame of the clip layer, given the current transient zoom scale, and the
    current frame of the clip layer. Animate the `position` and `bounds.size` of the clip layer
    using this rect.

2.  Reposition the scrolled contents layer under the clip layer, to ensure that content underneath
    shows up in the correct location after repositioning the clip layer position in (1).

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::removeTransientZoomFromLayer):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::adjustTransientZoom):

Plumb the `originInVisibleRect` alongside the existing `origin`, which is in `layerForPageScaling`&apos;s
coordinate space.

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):

Remove the position and size animations from the scrolled contents and clip layers when committing
the transient zoom, to prevent a jump in the case when we&apos;re rubber-banding against the min or max
scale before ending transient zooming.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::applyMagnification):

Pass in `m_magnificationOrigin` here; this will eventually become the drawing area&apos;s
`m_transientZoomOriginInVisibleRect`.

* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.h:
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::adjustTransientZoom):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::didCollectGeometryForSmartMagnificationGesture):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):

Set the `mainFrameClipLayerID` here.

Canonical link: <a href="https://commits.webkit.org/295495@main">https://commits.webkit.org/295495@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8955b1dbceb5988f0da36601a3f65d27f95c6c7b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105211 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24923 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110425 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55875 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107252 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25351 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33467 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79909 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19762 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94962 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60216 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/104689 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19519 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/13037 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55265 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89222 "Passed tests") | [![loading-orange](https://github-production-user-asset-6210df.s3.amazonaws.com/3098702/291015173-08c448be-ac0a-4fd6-92a3-8165057445b7.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13081 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4; Running apply-patch; Checked out pull request; Running run-layout-tests-in-stress-mode; 14 flakes 2 failures; Uploaded test results; 10 flakes") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/113008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32373 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23848 "Passed tests") | [⏳ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/GTK-WK2-Tests-EWS "Waiting to run tests") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32738 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91179 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88619 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22608 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33515 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11303 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27784 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32297 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37711 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32081 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35424 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33645 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/b97ea8b926d6ce8fef4d94814a073ce165b95ecf. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45930

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/238a77b3d83ff12fd4454af69ae8ecfecb135f36. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45930

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -241,34 +247,61 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+    RetainPtr animation = fillFowardsAnimationWithKeyPath(keyPath);
+    [animation setFromValue:value];
+    [animation setToValue:value];
+    return animation.autorelease();
```
`.autorelease()` to return a `RetainPtr`?

whsieh replied:
Whoops! Yeah, that should just be `return animation;`

aprotyas commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
```diff
@@ -241,34 +247,61 @@ explicit RemoteLayerTreeDisplayLinkClient(WebPageProxyIdentifier pageID)
+
+    auto scaleForClipLayerAdjustment = std::max(1.0, *m_transientZoomScale);
+    auto clipLayerPosition = FloatPoint { [clipLayer position] };
+    auto clipLayerZoomOrigin = clipLayerPosition + *m_transientZoomOriginInVisibleRect;
```
Consider `ASSERT(m_transientZoomOriginInVisibleRect)`

whsieh replied:
Good call, will add this next to the extant `ASSERT(m_transientZoomOrigin);`.

Thanks for the review!

aprotyas reviewed (approved):
r+ just two small comments

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/cf50311cbc3f879d2fb8dc4b71d79f6fe6864b22. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45930

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a6cbe869acaa410b71899b217042778617de12c9. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45930

whsieh committed:
[macOS] Content under left obscured content inset area is clipped when pinch-zooming
https://bugs.webkit.org/show_bug.cgi?id=293600
rdar://151854427

Reviewed by Abrar Rahman Protyasha and Simon Fraser.

When applying transient zoom (i.e. during pinch zooming on macOS) while the left obscured content
inset is nonzero, web content underneath the obscured content inset area is clipped by the main
frame's clip layer, until the transient zoom finishes and the clip layer size and position is
updated to include the obscured content inset area.

This happens because transient zoom currently works by applying a transform to only the layer for
page scaling (i.e. the `RenderView`'s primary layer); however, since the page is contained within a
clip layer whose bounds don't change while in transient zoom state (during pinch zoom), any content
that would normally show up in the obscured inset area is clipped.

To fix this, we add logic to temporarily reposition and inflate the size of the clip layer (as well
as the scrolled contents layer underneath) while the user is in transient zoom, such that the clip
layer no longer clips any part of the page that falls inside a content inset area. See below for
more details.

* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::clipLayer const):

Add a helper method to grab the frame view's render layer compositor's top-level clip layer.

* Source/WebCore/page/LocalFrameView.h:
* Source/WebCore/platform/graphics/GeometryUtilities.cpp:
(WebCore::scaledRectAtOrigin):

Add a helper method that scales the given `FloatRect`, relative to the given `origin` point.

* Source/WebCore/platform/graphics/GeometryUtilities.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::mainFrameClipLayerID const):
(WebKit::RemoteLayerTreeTransaction::setMainFrameClipLayerID):

Add plumbing to provide the main frame clip layer ID in remote layer tree transactions, so that we
can obtain the clip layer in the UI-side remote drawing area proxy below.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.cpp:
(WebKit::DrawingAreaProxyCoordinatedGraphics::adjustTransientZoom):
* Source/WebKit/UIProcess/CoordinatedGraphics/DrawingAreaProxyCoordinatedGraphics.h:
* Source/WebKit/UIProcess/DrawingAreaProxy.h:
(WebKit::DrawingAreaProxy::adjustTransientZoom):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:

Rename the existing `m_transientZoomOrigin` to `m_transientZoomOriginInLayerForPageScale`, to
differentiate it from the new `m_transientZoomOriginInVisibleRect` (which is used to resize and
reposition the clip layer).

* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeDrawingAreaProxyMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::didCommitLayerTree):

Obtain the layer IDs of the main frame clip layer, as well as the scrolled contents layer.

(WebKit::fillFowardsAnimationWithKeyPathAndValue):
(WebKit::transientZoomTransformOverrideAnimation):
(WebKit::transientSizeAnimation):
(WebKit::transientPositionAnimation):

Add more helper functions here to create size (`bounds.size`) and position animations, which we
apply to the clipping and scrolled contents layers below.

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::applyTransientZoomToLayer):

Implement the main fix here:

1.  Compute the scaled frame of the clip layer, given the current transient zoom scale, and the
    current frame of the clip layer. Animate the `position` and `bounds.size` of the clip layer
    using this rect.

2.  Reposition the scrolled contents layer under the clip layer, to ensure that content underneath
    shows up in the correct location after repositioning the clip layer position in (1).

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::removeTransientZoomFromLayer):
(WebKit::RemoteLayerTreeDrawingAreaProxyMac::adjustTransientZoom):

Plumb the `originInVisibleRect` alongside the existing `origin`, which is in `layerForPageScaling`'s
coordinate space.

(WebKit::RemoteLayerTreeDrawingAreaProxyMac::commitTransientZoom):

Remove the position and size animations from the scrolled contents and clip layers when committing
the transient zoom, to prevent a jump in the case when we're rubber-banding against the min or max
scale before ending transient zooming.

* Source/WebKit/UIProcess/ViewGestureController.cpp:
(WebKit::ViewGestureController::applyMagnification):

Pass in `m_magnificationOrigin` here; this will eventually become the drawing area's
`m_transientZoomOriginInVisibleRect`.

* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.h:
* Source/WebKit/UIProcess/mac/TiledCoreAnimationDrawingAreaProxy.mm:
(WebKit::TiledCoreAnimationDrawingAreaProxy::adjustTransientZoom):
* Source/WebKit/UIProcess/mac/ViewGestureControllerMac.mm:
(WebKit::ViewGestureController::didCollectGeometryForSmartMagnificationGesture):
* Source/WebKit/WebProcess/WebPage/mac/RemoteLayerTreeDrawingAreaMac.mm:
(WebKit::RemoteLayerTreeDrawingAreaMac::willCommitLayerTree):

Set the `mainFrameClipLayerID` here.

Canonical link: https://commits.webkit.org/295495@main
